cpp/benchmarks/bench_utils.cc

```diff
@@ -0,0 +1,116 @@
+//  Copyright 2022 Lance Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#include "bench_utils.h"
+
+#include <arrow/dataset/discovery.h>
+#include <arrow/dataset/file_parquet.h>
+#include <arrow/io/api.h>
+#include <arrow/util/string.h>
+#include <fmt/format.h>
+#include <lance/arrow/file_lance.h>
+
+#include <filesystem>
+#include <future>
+
+std::shared_ptr<::arrow::dataset::Scanner> OpenScanner(
+    const std::string& uri,
+    const std::vector<std::string>& columns,
+    std::optional<arrow::compute::Expression> filter,
+    std::optional<int> batch_size) {
+  auto format = std::shared_ptr<arrow::dataset::FileFormat>();
+  if (uri.ends_with(".lance")) {
+    format = lance::arrow::LanceFileFormat::Make();
+  } else if (uri.ends_with(".parquet") || uri.ends_with(".parq")) {
+    format.reset(new arrow::dataset::ParquetFileFormat());
+  }
+  auto factory = ::arrow::dataset::FileSystemDatasetFactory::Make(
+                     uri, format, arrow::dataset::FileSystemFactoryOptions())
+                     .ValueOrDie();
+  auto dataset = factory->Finish().ValueOrDie();
+  // fmt::print("{} dataset: {} groups={}\n", format->type_name(), uri,
+  // dataset->schema()->ToString());
+  auto scan_builder = dataset->NewScan().ValueOrDie();
+  if (batch_size.has_value()) {
+    fmt::print("Setting batch size: {}\n", batch_size.value());
+    assert(scan_builder->BatchSize(batch_size.value()).ok());
+  }
+  if (filter.has_value()) {
+    assert(scan_builder->Filter(filter.value()).ok());
+  }
+  assert(scan_builder->Project(columns).ok());
+  assert(scan_builder->UseThreads().ok());
+  auto scanner = scan_builder->Finish().ValueOrDie();
+  /// TODO: somehow scanner builder does not honor batch size??
+  scanner->options()->batch_size = batch_size.value();
+  if (format->type_name() == "lance") {
+    /// set to 16 will crash EC2 instance.
+    scanner->options()->batch_readahead = 8;
+  }
+  return scan_builder->Finish().ValueOrDie();
+}
+
+std::shared_ptr<::arrow::io::RandomAccessFile> OpenUri(const std::string& uri, bool ignore_error) {
+  if (uri.starts_with("s3")) {
+    auto fs = arrow::fs::FileSystemFromUriOrPath(uri);
+    if (!fs.ok() && !ignore_error) {
+      fs.ValueOrDie();
```

Comment:
Not a big deal since it's benchmark utilities but might be better to just explicitly throw an exception here with the Result status code+message?